File: src/design-pattern/km-singleton.py
from typing import Dict, Optional
from threading import Lock


class Cart:
    __instance: Optional["Cart"] = None  # private static
    __lock: Lock = Lock()  # private static

    def __init__(self) -> None:
        self.name2cnt: Dict[str, int] = {}

    @classmethod
    def getInstance(cls):
        if cls.__instance is None:
            with cls.__lock:
                if cls.__instance is None:
                    cls.__instance = Cart()

        return cls.__instance

    def add(self, name: str, cnt: int):
        if name in self.name2cnt:
            self.name2cnt[name] += cnt
        else:
            self.name2cnt[name] = cnt

    def print(self):
        for name, cnt in self.name2cnt.items():
            print(name, cnt)


try:
    while True:
        name, cntStr = input().strip().split(" ")
        cnt = int(cntStr)
        cart = Cart.getInstance()
        cart.add(name, cnt)

except EOFError as _:
    cart = Cart.getInstance()
    cart.print()

File: src/design-pattern/km-singleton.js
import { createInterface } from "node:readline";

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

class Cart {
  /** @type {Cart} */ static #instance; // = new Cart();
  /** @type {Map<string, number>} */ #nameToCnt = new Map();

  static getInstance() {
    if (!Cart.#instance) {
      Cart.#instance = new Cart();
    }
    return Cart.#instance;
  }

  /**
   *
   * @param {string} name
   * @param {number} cnt
   */
  add(name, cnt) {
    Cart.#instance.#nameToCnt.set(
      name,
      (Cart.#instance.#nameToCnt.get(name) ?? 0) + cnt,
    );
  }

  print() {
    for (const [name, cnt] of Cart.#instance.#nameToCnt.entries()) {
      console.log(name, cnt);
    }
  }
}

rl.on("line", (line) => {
  const [name, cntStr] = line.trim().split(" ");
  const cnt = Number.parseInt(cntStr);
  const instance = Cart.getInstance();
  instance.add(name, cnt);
});

rl.on("pause", () => {
  const instance = Cart.getInstance();
  instance.print();
  rl.close();
});
